blocks/ifus/ifus.js

```diff
@@ -0,0 +1,248 @@
+    ], countrySelector);
+  });
+  countrySelector.disabled = false;
+
+  searchButton.disabled = false;
+  searchButton.addEventListener('click', async () => {
+    clear.forEach((value) => { value.options[0].selected = true; });
+    const assets = idSelector.value && countrySelector.value
+      ? assetFilter(idSelector, countrySelector) : [];
+    result.innerHTML = '';
+    createDomStructure([{ type: 'h4', textContent: await translate('ifuSearchTitle', 'Search results') }], result);
+    if (assets.length === 0) {
+      createDomStructure([{ type: 'div', children: [{ type: 'strong', textContent: await translate('ifuSearchNoResult', 'No result') }] }], result);
+    } else {
+      const map = new Map();
+      assets.forEach((asset) => {
+        if (map.has(asset.Title)) {
+          const links = map.get(asset.Title);
+          links.push(asset.URL);
+        } else {
+          map.set(asset.Title, [asset.URL]);
+        }
+      });
+      await Promise.all(Array.from(map).map(async ([key, value]) => {
+        createDomStructure([
+          {
+            type: 'div',
+            children: [
+              { type: 'h5', textContent: key },
+              {
+                type: 'div',
+                children: [
+                  {
+                    type: 'div',
+                    children:
+                      [
+                        {
+                          type: 'div',
+                          textContent: productCodes(idSelector),
+                          children:
+                            [
+                              { type: 'h6', position: 'prepend', textContent: await translate('ifuSearchProductCodes', 'Product Code(s)') },
+                            ],
+                        },
+                      ],
+                  },
+                  {
+                    type: 'div',
+                    textContent: countrySelector.value,
+                    children:
+                      [
+                        { type: 'h6', position: 'prepend', textContent: await translate('ifuSearchCountrySelected', 'Country Selected') },
+                      ],
+                  },
+                  {
+                    type: 'div',
+                    children: value.map(
+                      (link) => ({
+                        type: 'a', attributes: { href: link, target: 'blank' }, children: [{ type: 'button', textContent: link.substring(link.lastIndexOf('/') + 1) }],
+                      }),
+                    ),
+                  },
+                ],
+              }],
+          },
+        ], result);
+      }));
+    }
+    result.classList.remove('no-result');
+  });
+}
+
+async function populate(block) {
+  const resp = await fetch('/en/dj-test.json?limit=10000');
+  if (resp.ok) {
+    const json = await resp.json();
+    const eIFUCode = block.querySelector('#eIFU-code');
```

Comment:
you could reduce the number of vars by making it generic:

```
  const getSelectors = (prefix) => {
    return {
      code: block.querySelector(`#${prefix}-code`),
      country: block.querySelector(`#${prefix}-country`),
      search: block.querySelector(`#${prefix}-code-search`),
    };
  };

  const eIFUSelectors = getSelectors('eIFU');
  const udiSelectors = getSelectors('udi');
  const productSelectors = getSelectors('product');
```

you could then wrap `populateSearch` in a callback generator function:
```
  const createPopulateSearchCallback = (type, idProducer, assetFilter, productCodesMapper) => () => {
    populateSearch(
      type.code,
      type.country,
      type.search,
      result,
      idProducer,
      () => json.countries.data,
      assetFilter,
      productCodesMapper,
      [eIFUSelectors.code, eIFUSelectors.country, udiSelectors.code, udiSelectors.country],
    );
  };
```

later generate all your callbacks:
```
  const populateSearchCallbacks = [
    createPopulateSearchCallback(
      eIFUSelectors,
      () => json.eIFUs.data.map((value) => value.ID),
      (idSelector, countrySelector) => json.IFUAssets.data.filter(
        (asset) => asset.eIFU === idSelector.value && asset.Country === countrySelector.value,
      ),
      (idSelector) => json.eIFUToProductCodes.data
        .filter((match) => match.eIFU_ID === idSelector.value)
        .map((match) => match['Product Code']).join(', '),
    ),
    ...
]
```

and finally call the callbacks:
```
populateSearchCallbacks.forEach((callback) => callback());
```

nothing blocking, just more javascript'y - please ignore if stupid :)

Reply:
Yes, cleaning it up by creating a selectors object makes a lot of sense - actually, I did go a little further and added the rest of the functions as well and reuse the selectors array for the clean variable. In total, I think this makes the code a lot cleaner - thanks a lot!